File: aspnet-in-practice/05/CS/DynamicData/ASPNET4InPractice.DynamicData/ObjectModel/DataAnnotations.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel;

namespace ASPNET4InPractice.DynamicData.ObjectModel
{
	[MetadataType(typeof(CustomerMetaData))]
	public partial class Customer
	{ }

	[MetadataType(typeof(ProductMetadata))]
	public partial class Product
	{ }

	public class CustomerMetaData
	{
		[DataTypeAttribute(DataType.MultilineText)]
		public string Address { get; set; }

		[UIHintAttribute("Phone")]
		public string Phone { get; set; }
	}

	public class ProductMetadata
	{
		[DisplayFormat(ApplyFormatInEditMode = false,
					   DataFormatString = "{0:C}",
					   NullDisplayText = "not set")]
		[DisplayName("Price")]
		[Required]
  		[Range(0, 100, ErrorMessage="Valid only between 0 and 100")]
		public decimal UnitPrice { get; set; }
	}
}

File: aspnet-in-practice/05/VB/DynamicData/ASPNET4InPractice.DynamicData/ObjectModel/DataAnnotations.vb
Imports System
Imports System.ComponentModel.DataAnnotations
Imports System.ComponentModel

<MetadataType(GetType(CustomerMetaData))>
Partial Public Class Customer
End Class

<MetadataType(GetType(ProductMetadata))>
Partial Public Class Product
End Class

Public Class CustomerMetaData
	<DataTypeAttribute(DataType.MultilineText)>
	Public Property Address() As String

	<UIHintAttribute("Phone")>
	Public Property Phone() As String
End Class

Public Class ProductMetadata
	<DisplayFormat(ApplyFormatInEditMode:=False, DataFormatString:="{0:C}", NullDisplayText:="not set")>
	<DisplayName("Price")>
	<Required()>
	<Range(0, 100, ErrorMessage:="Valid only between 0 and 100")>
	Public Property UnitPrice() As Decimal
End Class